File: backend/app/features/twin/schemas/ai_activity.py
"""Schemas for AI Activity sessions proxied from detection-engine."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class AiSessionCounters(BaseModel):
    tool_executions: int = 0
    files_read: int = 0
    files_modified: int = 0
    network_connections: int = 0
    process_count: int = 0


class AiSessionSummary(BaseModel):
    session_id: str
    device_id: str
    app_name: str
    root_process_id: str
    root_pid: int = 0
    start_time: str
    end_time: Optional[str] = None
    duration_ms: Optional[int] = None
    status: str = "active"
    counters: AiSessionCounters = Field(default_factory=AiSessionCounters)
    external_domains: List[str] = Field(default_factory=list)
    secrets_accessed: List[str] = Field(default_factory=list)
    git_repos: List[str] = Field(default_factory=list)
    docker_activity: List[str] = Field(default_factory=list)
    k8s_activity: List[str] = Field(default_factory=list)
    cloud_activity: List[str] = Field(default_factory=list)
    risk_score: int = 0
    risk_factors: List[str] = Field(default_factory=list)
    finding_count: int = 0
    process_count: int = 0


class AiSessionListResponse(BaseModel):
    device_id: str
    total: int = 0
    items: List[AiSessionSummary] = Field(default_factory=list)


class AiSessionDetailResponse(AiSessionSummary):
    processes: Dict[str, Any] = Field(default_factory=dict)
    tools: List[Dict[str, Any]] = Field(default_factory=list)
    findings: List[Dict[str, Any]] = Field(default_factory=list)
    spawn_tree: Dict[str, Any] = Field(default_factory=dict)
    activity_chain: List[Dict[str, Any]] = Field(default_factory=list)


class AiSessionGraphResponse(BaseModel):
    session_id: str
    device_id: str
    app_name: str
    graph: Dict[str, Any] = Field(default_factory=dict)


class AiSessionTimelineResponse(BaseModel):
    session_id: str
    device_id: str
    total: int = 0
    items: List[Dict[str, Any]] = Field(default_factory=list)


class AiSessionChainResponse(BaseModel):
    session_id: str
    device_id: str
    app_name: str
    total: int = 0
    items: List[Dict[str, Any]] = Field(default_factory=list)


class AiProcessLookupResponse(BaseModel):
    process_id: str
    session_id: str
    device_id: str
    app_name: str
    role: str = "unknown"

File: frontend/src/features/twin/types/aiActivity.ts
export type AiSessionCounters = {
  tool_executions: number;
  files_read: number;
  files_modified: number;
  network_connections: number;
  process_count: number;
};

export type AiSessionSummary = {
  session_id: string;
  device_id: string;
  app_name: string;
  root_process_id: string;
  root_pid: number;
  start_time: string;
  end_time?: string | null;
  duration_ms?: number | null;
  status: string;
  counters: AiSessionCounters;
  external_domains: string[];
  secrets_accessed: string[];
  git_repos: string[];
  docker_activity: string[];
  k8s_activity: string[];
  cloud_activity: string[];
  risk_score: number;
  risk_factors: string[];
  finding_count: number;
  process_count: number;
};

export type AiSessionListResponse = {
  device_id: string;
  total: number;
  items: AiSessionSummary[];
};

export type AiSessionDetail = AiSessionSummary & {
  processes: Record<string, unknown>;
  tools: Array<Record<string, unknown>>;
  findings: Array<Record<string, unknown>>;
  spawn_tree: Record<string, unknown>;
  activity_chain?: AiActivityChainStep[];
};

export type AiActivityChainStep = {
  kind: string;
  label: string;
  detail?: string;
  timestamp?: string;
  count?: number;
  artifacts?: Record<string, unknown>;
};

export type AiSessionChainResponse = {
  session_id: string;
  device_id: string;
  app_name: string;
  total: number;
  items: AiActivityChainStep[];
};

export type AiSessionGraphResponse = {
  session_id: string;
  device_id: string;
  app_name: string;
  graph: {
    nodes: Array<{ id: string; label: string; kind: string; role?: string; meta?: Record<string, unknown> }>;
    edges: Array<{ source: string; target: string; relation: string; meta?: Record<string, unknown> }>;
  };
};

export type AiSessionTimelineResponse = {
  session_id: string;
  device_id: string;
  total: number;
  items: Array<{
    kind: string;
    timestamp: string;
    process_id: string;
    summary: string;
    artifacts?: Record<string, unknown>;
  }>;
};
